pipeline.go

```diff
@@ -51,11 +56,75 @@ func setup(path string) error {
 }
 
 //Run executes the pipeline
-func Run(pipeline Pipeline) ([]StageExecutionResult, error) {
-	if err := setup(pipeline.DefaultRepo); err != nil {
+func (p *Pipeline) Run() ([]StageExecutionResult, error) {
+	if err := setup(p.DefaultRepo); err != nil {
 		return nil, fmt.Errorf("error in inicial setup. %q", err)
+	}
+
+	for _, stage := range p.Stages {
+		var er StageExecutionResult
+		var err error
 
+		id := fmt.Sprintf("%s-%s-%s", stage.Name, p.DefaultEnv["month"], p.DefaultEnv["year"])
+		log.Printf("Executing %s ...\n", id)
+
+		er, err = build(stage)
+		if err != nil {
+			return nil, fmt.Errorf("error in image build %s", err)
+		}
+
+		fmt.Println(er)
 	}
 
 	return []StageExecutionResult{}, nil
 }
+
+// Build tries to build a docker image for a job and panics if it can not suceed.
+func build(s Stage) (StageExecutionResult, error) {
+	id := fmt.Sprintf("%s", s.Name)
+	log.Printf("Building image %s...\n", id)
+
+	pi, err := buildImage(s.Dir, s.Env["commit"])
+	if err != nil {
+		return *pi, fmt.Errorf("Error building DataCollector image %s: %q", id, err)
+	} else if status.Code(pi.ExitStatus) != status.OK {
+		return *pi, fmt.Errorf("Status code %d(%s) building DataCollector image %s", pi.ExitStatus, status.Text(status.Code(pi.ExitStatus)), id)
+	}
+	log.Printf("Image %s build sucessfully\n", id)
+	return *pi, nil
+}
+
+// build runs a go build for each path. It will also insert the value of main.gitCommit in the binaries.
+func buildImage(dir, commit string) (*StageExecutionResult, error) {
+	cmdList := strings.Split(fmt.Sprintf("docker build --build-arg GIT_COMMIT=%s -t %s .", commit, filepath.Base(dir)), " ")
```

Comment:
Aqui temos uma variável hardcoded, isso também não pode estar aqui. Uma solução simples é passar todas as variáveis de ambiente como argumentos do build;

- [Tutorial de concatenação de string em Go (bytes.Buffer for the win!)](https://yourbasic.org/golang/build-append-concatenate-strings-efficiently/)
- [Docker build command line](https://docs.docker.com/engine/reference/commandline/build/)

Reply:
O ajuste foi realizado.